### PR TITLE
Move ComHost generation from dotnet/sdk to Microsoft.NET.HostModel

### PR DESCRIPTION
**Pull request by jkoritzinsky**
For .NET Core 3.0, we added the support classes for generating the ClsidMap, embedding the ClsidMap into the COM Host, and generating the RegFreeComManifest to dotnet/sdk for simplicity. At the end of the 3.0 cycle, we created the Microsoft.NET.HostModel package for the single-file bundler. This PR moves the COM Host code into Microsoft.NET.HostModel.

Additionally, this PR also adds many tests unit tests for the ClsidMap and the Reg Free COM Manifest generation that would have been very annoying to add into the dotnet/sdk test suite.

Contributes to #6338

cc: @nguerrera

### CODE REVIEW

File: src/managed/Microsoft.NET.HostModel/ComHost/ComHost.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+            var destinationDirectory = new FileInfo(comHostDestinationFilePath).Directory.FullName;
+            if (!Directory.Exists(destinationDirectory))
+            {
+                Directory.CreateDirectory(destinationDirectory);
+            }
+
+            // Copy apphost to destination path so it inherits the same attributes/permissions.
+            File.Copy(comHostSourceFilePath, comHostDestinationFilePath, overwrite: true);
+
+            if (ResourceUpdater.IsSupportedOS())
```
Comment (AaronRobinsonMSFT) [inline src/managed/Microsoft.NET.HostModel/ComHost/ComHost.cs]: Flip this predicate and simply throw an exception. The current contents should be moved to the end of the function no need for an `else` clause.

File: src/managed/Microsoft.NET.HostModel/ComHost/ConflictingGuidException.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.NET.HostModel.ComHost
+{
+    public class ConflictingGuidException : Exception
```
Comment (AaronRobinsonMSFT) [inline src/managed/Microsoft.NET.HostModel/ComHost/ConflictingGuidException.cs]: Do we need to document this class?

File: src/managed/Microsoft.NET.HostModel/ComHost/MissingGuidException.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.NET.HostModel.ComHost
+{
+    public class MissingGuidException : Exception
+    {
+        public MissingGuidException(string typeName)
+        {
+            TypeName = typeName;
```
Comment (AaronRobinsonMSFT) [inline src/managed/Microsoft.NET.HostModel/ComHost/MissingGuidException.cs]: Assert non-null?

File: src/managed/Microsoft.NET.HostModel/ComHost/ConflictingGuidException.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        public ConflictingGuidException(string typeName1, string typeName2, Guid guid)
+        {
+            TypeName1 = typeName1;
```
Comment (AaronRobinsonMSFT) [inline src/managed/Microsoft.NET.HostModel/ComHost/ConflictingGuidException.cs]: Assert non-null?

File: src/test/Assets/TestProjects/ComLibraryConflictingGuid/ComLibraryConflictingGuid.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+using System;
+using System.Runtime.InteropServices;
+
+namespace ComLibrary
+{
+    [ComVisible(true)]
+    [Guid("cc6e9910-18d5-484a-a2d2-fa8910fd0261")]
```
Comment (AaronRobinsonMSFT) [inline src/test/Assets/TestProjects/ComLibraryConflictingGuid/ComLibraryConflictingGuid.cs]: Please comment on these classes that the guid should be the same. I realize the name of the class is indicating this.

File: src/test/Microsoft.NET.HostModel.Tests/Microsoft.NET.HostModel.ComHost.Tests/ClsidMapTests.cs [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                {
+                    if (peReader.HasMetadata)
+                    {
+                        MetadataReader reader = peReader.GetMetadataReader();
+                        ClsidMap.Create(reader, clsidMapPath);
+                    }
+                }
+            }
+
+            using var clsidMapFile = File.OpenText(clsidMapPath);
```
Comment (AaronRobinsonMSFT) [inline src/test/Microsoft.NET.HostModel.Tests/Microsoft.NET.HostModel.ComHost.Tests/ClsidMapTests.cs]: Please use at least one level of brace scoping here. this is really terse for no real reason.

Comment (jkoritzinsky) [reply src/test/Microsoft.NET.HostModel.Tests/Microsoft.NET.HostModel.ComHost.Tests/ClsidMapTests.cs]: Yeah the new VS code helpers for the `using statement` feature are really aggressive. I'll switch it back to the traditional style.

Comment (AaronRobinsonMSFT) [reply src/test/Microsoft.NET.HostModel.Tests/Microsoft.NET.HostModel.ComHost.Tests/ClsidMapTests.cs]: Thanks!

File: src/managed/Microsoft.NET.HostModel/ComHost/ClsidMap.cs [hunk 7]
```diff
@@ -0,0 +1,282 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
Comment (jkotas) [inline src/managed/Microsoft.NET.HostModel/ComHost/ClsidMap.cs]: Non-standard license header

File: src/managed/Microsoft.NET.HostModel/ComHost/ComHost.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
+﻿using System;
```
Comment (jkotas) [inline src/managed/Microsoft.NET.HostModel/ComHost/ComHost.cs]: Missing license header

Comment (jkotas) [reply src/managed/Microsoft.NET.HostModel/ComHost/ComHost.cs]: (multiple-files)

**Pull request merged**